### PR TITLE
Core allocation/reservation logic renovation

### PR DESCRIPTION
**Pull request by yangwwei**

### CODE REVIEW

**Comment by yangwwei** [comment]
Hi @wilfred-s 

Thanks for the patch, overall the patch does a really good job to clean up things. Now the structure looks very clear. I have some high-level comments, most about reservation code:

1. scheduling_application: 392-398 this internal unreserve -- would it cause some left-over things? compare to scheduling_partiton#unReserve. Ideally, it will be great if we can encapsulate the unreserve logic in functions and just call them when needed.

2. We need to make sure in `tryAllocate`, when an allocation is made for a reserved request, we need to do unreserve properly.  Right now, it seems the unreserve did not happen for this case.

3. For reserved apps, if allocation from reservation doesn't work, let's try to allocate on all nodes if possible.  Otherwise, such reservations may be "stuck" on some nodes which doesn't serve the reservation purpose.

I will look into more details shortly.

Thanks

**Comment by yangwwei** [comment]
The other thing is, when we score nodes while attempting to reserve resources, we need to make sure the candidate node satisfies the following condition
- the node fits the predicates
- the node has not been reserved
- the node is scheduable

right now the score is calculated based on resource only, that will lead the issue that reserve resources on a node that might not have a certain label to satisfy a node-selector, or against a affinity/anti-affinity constraint. which is an invalid reservation.

**Comment by wilfred-s** [comment]
New commits pushed with the smoke tests and further clean up. To link this back to the comment from @yangwwei:
comment 1: (https://github.com/apache/incubator-yunikorn-core/pull/89#issuecomment-584913754) remarks:
1.  internal unreserve: fixed the issue found (commit 1)
1.  tryAllocate correctly unreserves (commit 2)
1. if reserved allocation fails try all nodes (commit 1)

comment 2 (https://github.com/apache/incubator-yunikorn-core/pull/89#issuecomment-585361646) remarks
1. running predicates for each allocation try for each  node in all cycles will cause a huge slow down. For example in a 100 node cluster 1 predicates check is run if there are enough resources available on the node. If we do it before that check and let it lead us and we have 99 nodes that do not fit the ask we would have run the predicates 100 times for the same alloc. Caching the predicate run is not possible as node usage can change and thus the predicate outcome would change. I think that is 1) is thus a no go.
1. the score used is really basic at the moment. However I could argue for or against all scores. A large node might have a longer average runtime per allocation (service type load) and thus release less often. Without metrics we really cannot argue for one or the other or for a 3rd alternative.
1. Yes we need better metrics, I will follow up with a new jira

For the test failures: I have seen a number of them and they are transient. The tests use a manual scheduler (steps based on a counter). The manual scheduling in the smoke tests is I think the cause of the issue. The duration of the scheduling cycle is short and also cut even shorter when nothing needs to be done. We probably _waste_ scheduling cycles because we have nothing to do. When events are later processed we have no scheduling cycles to progress.
I am thinking about a better solution or even using continuous scheduling in the smoke tests.

File: pkg/scheduler/scheduling_application.go [hunk 1]
```diff
@@ -19,22 +19,604 @@
...
+
+// Return the pending resources for this application
+func (sa *SchedulingApplication) GetPendingResource() *resources.Resource {
+	sa.RLock()
+	defer sa.RUnlock()
+	return sa.pending
+}
+
+// Return the allocating and allocated resources for this application
+func (sa *SchedulingApplication) getUnconfirmedAllocated() *resources.Resource {
```
Comment (yangwwei) [inline pkg/scheduler/scheduling_application.go]: allocated is confirmed resource
here it calls (allocated + allocating = unconfirmedAllocated) is a bit misleading. Can we rename this to `getAssumeAllocated`?

File: pkg/cache/initializer.go [hunk 2]
```diff
@@ -194,21 +194,16 @@ func checkResourceConfigurationsForQueue(cur *QueueInfo, parent *QueueInfo) erro
 		cur.GuaranteedResource = resources.NewResource()
 	}
 
-	// If Max resource not set, use parent's max resource
-	if cur.MaxResource == nil && parent != nil && parent.MaxResource != nil {
-		cur.MaxResource = parent.MaxResource
-	}
-
```
[the remark this answers is not in the document]
Comment (wilfred-s) [reply pkg/cache/initializer.go]: yes, max resources is calculated recursively automatically in the new logic

**Comment by yangwwei** [comment]
LGTM, @wilfred-s  pls merge with a proper commit message.
Let's track the flaky tests in a separate issue.

**Pull request merged**